src/app/modules/main/communities/module.nim

```diff
@@ -531,13 +531,71 @@ method callbackFromAuthentication*(self: Module, authenticated: bool) =
 method getCommunityPublicKeyFromPrivateKey*(self: Module, communityPrivateKey: string): string =
   result = self.controller.getCommunityPublicKeyFromPrivateKey(communityPrivateKey)
 
+method updateTokenModelForCommunity*(self: Module, communityId: string, sharedAddresses: seq[string]) =
+  self.controller.asyncCheckPermissionsToJoin(communityId, sharedAddresses)
+  self.controller.asyncCheckAllChannelsPermissions(communityId, sharedAddresses)
+
 method prepareTokenModelForCommunity*(self: Module, communityId: string) =
   let community = self.controller.getCommunityById(communityId)
   var tokenPermissionsItems: seq[TokenPermissionItem] = @[]
 
   for id, tokenPermission in community.tokenPermissions:
-    let chats = self.controller.getChatDetailsByIds(tokenPermission.chatIDs)
+    let chats = community.getCommunityChats(tokenPermission.chatIds)
     let tokenPermissionItem = buildTokenPermissionItem(tokenPermission, chats)
     tokenPermissionsItems.add(tokenPermissionItem)
 
-  self.view.spectatedCommunityPermissionModel.setItems(tokenPermissionsItems)
\ No newline at end of file
+  self.view.spectatedCommunityPermissionModel.setItems(tokenPermissionsItems)
+  self.updateTokenModelForCommunity(communityId, @[])
+
+proc applyPermissionResponse*(self: Module, communityId: string, permissions: Table[string, CheckPermissionsResultDto]) =
+  let community = self.controller.getCommunityById(communityId)
+  for id, criteriaResult in permissions:
+    if community.tokenPermissions.hasKey(id):
+      let tokenPermissionItem = self.view.spectatedCommunityPermissionModel.getItemById(id)
+      if tokenPermissionItem.id == "":
+        continue
+
+      var updatedTokenCriteriaItems: seq[TokenCriteriaItem] = @[]
+      var permissionSatisfied = true
+
+      for index, tokenCriteriaItem in tokenPermissionItem.getTokenCriteria().getItems():
+
+        let updatedTokenCriteriaItem = initTokenCriteriaItem(
+          tokenCriteriaItem.symbol,
+          tokenCriteriaItem.name,
+          tokenCriteriaItem.amount,
+          tokenCriteriaItem.`type`,
+          tokenCriteriaItem.ensPattern,
+          criteriaResult.criteria[index]
+        )
+
+        if criteriaResult.criteria[index] == false:
+          permissionSatisfied = false
```

Comment:
This seems wrong. Permission is satisfied if any criteria is met :thinking: Have you tested a permission with more than one criteria?

Reply:
This code was just copied by Boris from the chat_section module to the communities module, so I'm not 100% sure of what everything does.

I think this is right, criterias act as logical ANDs. Having multiple permissions acts as `OR`.
(I had to check the design as well to refresh my memory :smile: https://www.figma.com/file/17fc13UBFvInrLgNUKJJg5/Kuba%E2%8E%9CDesktop?type=design&node-id=2367-280302&mode=design&t=KT4mjhyWRIJbGmUs-0)

Reply:
Ah, right, sorry for confusion :pray: